backend/groth16/bn254/setup_gpu.go

```diff
@@ -0,0 +1,810 @@
+	g1Scalars = append(g1Scalars, toxicWaste.alpha, toxicWaste.beta, toxicWaste.delta)
+	g1Scalars = append(g1Scalars, A...)
+	g1Scalars = append(g1Scalars, B...)
+	g1Scalars = append(g1Scalars, Z...)
+	g1Scalars = append(g1Scalars, vkK...)
+	g1Scalars = append(g1Scalars, pkK...)
+	for i := range ckK {
+		g1Scalars = append(g1Scalars, ckK[i]...)
+	}
+
+	g1PointsAff := curve.BatchScalarMultiplicationG1(&g1, g1Scalars)
+
+	// sets pk: [α]₁, [β]₁, [δ]₁
+	pk.G1.Alpha = g1PointsAff[0]
+	pk.G1.Beta = g1PointsAff[1]
+	pk.G1.Delta = g1PointsAff[2]
+
+	offset := 3
+	pk.G1.A = g1PointsAff[offset : offset+len(A)]
+	offset += len(A)
+
+	pk.G1.B = g1PointsAff[offset : offset+len(B)]
+	offset += len(B)
+
+	bitReverse(g1PointsAff[offset : offset+int(domain.Cardinality)])
+	sizeZ := int(domain.Cardinality) - 1 // deg(H)=deg(A*B-C/X^n-1)=(n-1)+(n-1)-n=n-2
+	pk.G1.Z = g1PointsAff[offset : offset+sizeZ]
+
+	offset += int(domain.Cardinality)
+
+	vk.G1.K = g1PointsAff[offset : offset+nbPublicWires]
+	offset += nbPublicWires
+
+	pk.G1.K = g1PointsAff[offset : offset+nbPrivateWires]
+	offset += nbPrivateWires
+
+	// ---------------------------------------------------------------------------------------------
+	// Commitment setup
+
+	commitmentBases := make([][]curve.G1Affine, len(commitmentInfo))
+	for i := range commitmentBases {
+		size := len(ckK[i])
+		commitmentBases[i] = g1PointsAff[offset : offset+size]
+		offset += size
+	}
+	if offset != len(g1PointsAff) {
+		return errors.New("didn't consume all G1 points") // TODO @Tabaie Remove this
+	}
+
+	pk.CommitmentKeys, vk.CommitmentKey, err = pedersen.Setup(commitmentBases...)
+	if err != nil {
+		return err
+	}
+
+	vk.PublicAndCommitmentCommitted = commitmentInfo.GetPublicAndCommitmentCommitted(commitmentWires, r1cs.GetNbPublicVariables())
+
+	// ---------------------------------------------------------------------------------------------
+	// G2 scalars
+
+	// the G2 scalars are ordered as follow:
+	//
+	// [[B(i)], [β], [δ], [γ]]
+	// len(B) == nbWires
+
+	// compute our batch scalar multiplication with g2 elements
+	g2Scalars := append(B, toxicWaste.beta, toxicWaste.delta, toxicWaste.gamma)
+
+	g2PointsAff := curve.BatchScalarMultiplicationG2(&g2, g2Scalars)
+
+	pk.G2.B = g2PointsAff[:len(B)]
+
+	// sets pk: [β]₂, [δ]₂
+	pk.G2.Beta = g2PointsAff[len(B)+0]
+	pk.G2.Delta = g2PointsAff[len(B)+1]
+
+	// sets vk: [δ]₂, [γ]₂
+	vk.G2.Delta = g2PointsAff[len(B)+1]
+	vk.G2.Gamma = g2PointsAff[len(B)+2]
+
+	// ---------------------------------------------------------------------------------------------
+	// Pairing: vk.e
+	vk.G1.Alpha = pk.G1.Alpha
+	vk.G2.Beta = pk.G2.Beta
+
+	// unused, here for compatibility purposes
+	vk.G1.Beta = pk.G1.Beta
+	vk.G1.Delta = pk.G1.Delta
+
+	if err := vk.Precompute(); err != nil {
+		return err
+	}
+
+	// set domain
+	pk.Domain = *domain
+
+	// Move static values (points, domain, hpoly denom) to the device/GPU
+	err = pk.setupDevicePointers()
```

Comment:
We have changed the ProvingKey struct to contain pointers to GPU memory so we would need a new structure to retain these pointers. I guess we can move this to `prove_gpu.go` but since we need to retain a reference to the struct we would also need to create a function that populates the struct and then pass this new struct into Prove to be used